Added method for combining path strings

King2500 opened the pull request:
As mentioned in #3 the .NET `Path` class has a few helpful methods; one still missing is a `combine` method.

Developers often need to combine path strings or partial paths. In most cases the base path is variable (maybe from a config file or database), so you dont know if it has a trailing "/" at the end and how to correctly append the subdir you're targeting:

```php
$newPath = $basePath.'subdir'; // does $basePath include a '/' at the end?
```

The .NET `Path` class has a `Combine` method, similar to what I suggest and implemented here.

```php
$newPath = Path::combine($basePath, 'subdir'); // it doesn't matter if $basePath 
                                               // includes the '/' at the end
```
Result is:
`/some/base/subdir`

To properly support relative paths, the resulting path gets canonicalized:
```php
// $basePath = '/some/base/';
$newPath = Path::combine($basePath, '../other');
```
=> `/some/other`

Supported are variable arguments and paths as array:
```php
Path::combine('/path', 'to', 'dir');  // => /path/to/dir
$paths = array('/path', 'to', 'dir');
Path::combine($paths);  // => /path/to/dir
```

webmozart commented:
Hi! :) At first glance, I don't see much of a difference to `Path::makeAbsolute()`. The example you gave can already be achieved with:

```php
$newPath = Path::makeAbsolute('subdir', $basePath);
```

As for combining multiple path parts, is there any use case where a simple `implode()` wouldn't do?

King2500 commented:
Using `Path::makeAbsolute` you need to need know (and make sure) `$basePath` is an absolute path. If it comes from a config file or database you maybe don't know about that. (Maybe it's configured by someone else). The order may also be confusing.

```
combine( parent, sub )  VS  
makeAbsolute( sub, parent )
```

Using `Path::makeRelative` for relative base paths produces a different result:

```php
$basePath = '../some/base';

echo Path::combine($basePath, 'subdir'); // => ../some/base/subdir
echo Path::makeRelative('subdir', $baseDir); // => ../../../subdir

// This throws an exception with above basePath:
Path::makeAbsolute('subdir', $basePath);
```

You are right about using `implode()` for simple concatenations, but if your parts include slashes or relative paths, you need to call `Path::canonicalize` afterwards. `Path::combine` simplifies everything. 

```php
Path::canonicalize(implode($basePath

King2500 commented:
Using `Path::makeAbsolute` you need to know (and make sure) `$basePath` is an absolute path. If it comes from a config file or database you maybe don't know about that. (Maybe it's configured by someone else). The argument order may also be confusing.

```
combine( parent, sub )  VS  makeAbsolute( sub, parent )
```

Using `Path::makeRelative` for relative base paths produces a different result:

```php
$basePath = '../some/base';

echo Path::combine($basePath, 'subdir'); // => ../some/base/subdir
echo Path::makeRelative('subdir', $baseDir); // => ../../../subdir

// This throws an exception with above basePath:
Path::makeAbsolute('subdir', $basePath);
```

You are right about using `implode()` for simple concatenations. But if your parts include slashes or relative paths, you need to call `Path::canonicalize` afterwards. `Path::combine` simplifies everything. 

```php
$basePath = '/some/base';
Path::canonicalize( implode('/', array($basePath, '../subdir', 'foo')) );
Path::combine($basePath, '../subdir', 'foo');
// both => /some/subdir/foo
```

I think it helps, because we often need to combine paths and everytime I do this I have to think about slashes. This is just my opinion. It's up to you, if you merge it or not. ;-)

webmozart commented:
Thank you for your detailed explanation :) This makes a lot of sense now.

One more question: The idea of `Path::makeAbsolute()` is to simplify the configuration of paths. The base path is supposed to be calculated by the application, while the other path is given by the user. Hence the user may both provide a relative or an absolute path:

```php
$pathFromConfig = getPathFromConfig();

// e.g. "../plugins" or "/path/to/plugins"

$absPath = Path::makeAbsolute($pathFromConfig, __DIR__.'/..');
```

This benefit is lost when using `Path::combine()`, because then the second path may not be absolute anymore:

```php
// works
$absPath = Path::combine(__DIR__.'/..', '../plugins', 'plugin');

// breaks
$absPath = Path::combine(__DIR__.'/..', '/path/to/plugins', 'plugin');
```

With `Path::makeAbsolute()`, however, you can easily solve the use case:

```php
$absPath = Path::makeAbsolute($pluginsDir.'/'.$plugin, __DIR__.'/..');
```

Since `Path::makeAbsolute()` canonicalizes the result, both `$pluginsDir` and `$plugin` may contain relative paths. Also leading/trailing slashes in both variables are ignored. At last, `$pluginsDir` may be both absolute and relative, which is not possible with `Path::combine()`. Hence I'm still not convinced that we really should add this method.

What do you think?

webmozart commented:
Ping.

King2500 commented:
Sorry, a little busy at the moment. I'll answer in the coming days.

King2500 commented:
Ok. Now my opinion on this:

I see two completely different purposes in this case, and so two methods.

**`makeAbsolute`** is used when you want to ensure you get an absolute path in the end, regardless if you have or have not an absolute path already. When the first parameter comes from a (user) configuration, it allows for absolute paths. Your base path MUST be absolute or it doesnt make sense.

**`combine`** is used when you dont care about absolute or relative (base) paths. You may have a base path as first parameter which may even come from a (system) configuration. But maybe its just the current path or root path. The main purpose is, you want to combine your (base) path (absolute or relative) with further path parts, which are intended as subpaths in most cases. These further parts may come from a (user) configuration and can ONLY be relative from this point. Because we want append/extend your base paths, not overwrite it with absolute (and so maybe completely different) paths. Your users can not inject absolute paths. (Ok, you can use ../ to escape, but thats not the purpose of a method called `combine`. You can check that with `isBasePath` then, which is meant for that special case). So your "breaking" example just doesnt make sense here.

**So in conclusion:**
* The (only) purpose of `makeAbsolute`is just to make sure a path is absolute, not combining path parts. (Same for `makeRelative` in the other way)
* The (only) purpose of `combine` is concatenating path parts in the sense of path strings, so you get a (normalized) path string in the end. (No need to think about absolute or relative paths here)

**Additional note:**
A further reason the method `Combine` exists in the .NET `Path` class, is the fact that you dont need to care about the path separator of the underlying system environment. You can just combine names and you get the correct `/` or `\` depending on the current OS:
```csharp
var path = Path.Combine(basePath, "plugins", name);
// "C:\basePath\plugins\name" on Windows
// "/basePath/plugins/name" on Unix
```
That's not that important in PHP but you eliminate a little more dependency on the environment and make it easier for the programmer (not) thinking about trailing/leading slashes and which slash to use.

I hope you get my point of view.

King2500 commented:
@webmozart 
Have you reflected / thought about my last post yet?

Here is a similar method in Ouzo framework, I've found:
https://github.com/letsdrink/ouzo/blob/master/src/Ouzo/Goodies/Utilities/Path.php#L29

King2500 commented:
Bump :)

webmozart commented on src/Path.php:
```diff
@@ -648,6 +648,48 @@ public static function getLongestCommonBasePath(array $paths)
+        $finalPath = null;
+
+        foreach ($paths as $path) {
+            if (null === $path || strlen((string)$path) == 0) {
```
Could you add a space after `(string)` here and in the other occurrences?

webmozart commented on src/Path.php:
```diff
@@ -648,6 +648,48 @@ public static function getLongestCommonBasePath(array $paths)
+     *
+     * @return null|string The combined path
+     *
+     * @since 1.1
```
Please change this to 1.2

webmozart commented on src/Path.php:
```diff
@@ -648,6 +648,48 @@ public static function getLongestCommonBasePath(array $paths)
+        }
+
+        if (null === $finalPath)
+            return null;
```
Please add braces for the if statement.

webmozart commented on tests/PathTest.php:
```diff
@@ -822,4 +822,80 @@ public function testIsBasePath($path, $ofPath, $result)
+            array('C:\\path\\to\\test', 'subdir', 'C:/path/to/test/subdir'),
+            array('C:\\path\\to\\test\\', 'subdir', 'C:/path/to/test/subdir'),
+            array('C:\\path\\to\\test', '/subdir', 'C:/path/to/test/subdir'),
+            array('C:\\path\\to\\test\\', '/subdir', 'C:/path/to/test/subdir'),
```
1.2 adds support for schemes (e.g. `phar://C:/foo/bar`, `phar:///foo/bar`, ...). Could you rebase this PR and add corresponding test cases?

webmozart replied:
I think you should also add test cases with roots only:

```
array('/', 'subdir', ...),
array('/', '/subdir', ...)
array('C:/', 'subdir', ...),
array('C:/', '/subdir', ...)
array('C:\\', 'subdir', ...),
array('C:\\', '/subdir', ...)
array('C:', 'subdir', ...),
array('C:', '/subdir', ...)
```

I think that the current code does not work yet with partition drives followed by a backslash.

webmozart commented:
Hi @King2500, thanks for your patience. In fact your last comment convinced me already, but this dropped below my sight. Could you make the changes that I suggested? After that, I'm :+1:

King2500 commented:
Rebased and changed.

Do you want me to combine the commits to one?

webmozart commented on src/Path.php:
```diff
@@ -681,6 +681,49 @@ public static function getLongestCommonBasePath(array $paths)
+        $finalPath = null;
+
+        foreach ($paths as $path) {
+            if (null === $path || strlen((string) $path) == 0) {
```
You can simplify this condition to

```
if ('' === (string) $path) {
```
```

webmozart commented on src/Path.php:
```diff
@@ -681,6 +681,49 @@ public static function getLongestCommonBasePath(array $paths)
+
+        $finalPath = null;
+
+        foreach ($paths as $path) {
```
I think it's better to do the string cast only once in the loop:

```
$path = (string) $path;
```

webmozart commented:
Yes, please squash the commits when you're done.

King2500 commented:
Following failed previously:
```php
array('C:/', '/subdir', 'C:/subdir'),
array('C:\\', '/subdir', 'C:/subdir'),
```
```
Expected : C:/subdir
Actual   : C://subdir
```
With last commit it is fixed. Can you review the last commit?
When you think it's OK, I'll squash all commits again and we finally come to an end / merge :)

King2500 commented:
We had a further bug. This failed after my previous commit:
```php
array('phar://', '/path/to/test', 'phar:///path/to/test'),
```
Fixed this and added more tests, plus some comments :)
Please review.

webmozart commented on tests/PathTest.php:
```diff
@@ -945,4 +945,106 @@ public function testIsBasePath($path, $ofPath, $result)
+            array(array('/path/to/test', 'subdir'), null, '/path/to/test/subdir'),
+            array(array('/path/to/test', '/subdir'), null, '/path/to/test/subdir'),
+            array(array('/path/to/test/', 'subdir'), null, '/path/to/test/subdir'),
+            array(array('/path/to/test/', '/subdir'), null, '/path/to/test/subdir'),
```
What happens when you pass an array with just one/no element? I think you should add test cases for that.

webmozart commented on tests/PathTest.php:
```diff
@@ -945,4 +945,106 @@ public function testIsBasePath($path, $ofPath, $result)
+            array('C:', 'subdir', 'C:/subdir'),
+            array('C:', '/subdir', 'C:/subdir'),
+
+            array('phar://', '/path/to/test', 'phar:///path/to/test'),
```
Missing test case:

```
array('phar:///', '/path/to/test', 'phar:///path/to/test'),
```

webmozart commented on tests/PathTest.php:
```diff
@@ -945,4 +945,106 @@ public function testIsBasePath($path, $ofPath, $result)
+            array('phar:///path/to/test/', '/subdir', 'phar:///path/to/test/subdir'),
+
+            array('phar://', 'C:/path/to/test', 'phar://C:/path/to/test'),
+            array('phar://', 'C:\\path\\to\\test', 'phar://C:/path/to/test'),
```
Missing test cases:

```
array('phar://C:', 'path/to/test', 'phar://C:/path/to/test'),
array('phar://C:', '/path/to/test', 'phar://C:/path/to/test'),
array('phar://C:/', 'path/to/test', 'phar://C:/path/to/test'),
array('phar://C:/', '/path/to/test', 'phar://C:/path/to/test'),
```

King2500 commented:
We already had a testcase with just one array element:
```php
array(array('/path'), null, '/path'),
```
Added all other tests.

Squash now?

webmozart commented:
Yes, looks good! :+1:

King2500 commented:
Ready :smirk_cat:

webmozart commented:
One last thing: You should add yourself as author to the class doc comment :)

webmozart commented:
Thanks @King2500!
